# Data Cleaning: New York City Airbnb Open Data

**Objective:**

> In this data cleaning project, I will clean the dataset by examining **missing values**, **duplicates**, **invalid/erroneous values**, fixing **datatypes**, **anomaly detection** and Y-variable cleaning methods (**Label encoding**, **dictionary mapping**, **one-hot encoding**).


<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#1.-Import-libraries" data-toc-modified-id="1.-Import-libraries-0">1. Import libraries</a></span></li><li><span><a href="#2.-Load-data-into-a-dataframe" data-toc-modified-id="2.-Load-data-into-a-dataframe-1">2. Load data into a dataframe</a></span></li><li><span><a href="#3.-Check-datatypes-within-the-data" data-toc-modified-id="3.-Check-datatypes-within-the-data-2">3. Check datatypes within the data</a></span></li><li><span><a href="#4.-Missing-Values" data-toc-modified-id="4.-Missing-Values-3">4. Missing Values</a></span></li><li><span><a href="#5.-Remove-NaN-values-from-name-and-host_name" data-toc-modified-id="5.-Remove-NaN-values-from-name-and-host_name-4">5. Remove NaN values from name and host_name</a></span></li><li><span><a href="#6.-Check-for-duplicate-values" data-toc-modified-id="6.-Check-for-duplicate-values-5">6. Check for duplicate values</a></span></li><li><span><a href="#7.-Remove-any-invalid-values" data-toc-modified-id="7.-Remove-any-invalid-values-6">7. Remove any invalid values</a></span></li><li><span><a href="#8.-Convert-NaN-reviews-per-month-and-NaT-last-reviews-into-'No-Reviews'" data-toc-modified-id="8.-Convert-NaN-reviews-per-month-and-NaT-last-reviews-into-'No-Reviews'-7">8. Convert NaN reviews per month and NaT last reviews into 'No Reviews'</a></span></li><li><span><a href="#9.-Convert-Latitude-and-Longitude-to-5-decimal-places" data-toc-modified-id="9.-Convert-Latitude-and-Longitude-to-5-decimal-places-8">9. Convert Latitude and Longitude to 5 decimal places</a></span></li><li><span><a href="#10.-Read-dataframe-to-CSV-file" data-toc-modified-id="10.-Read-dataframe-to-CSV-file-9">10. Read dataframe to CSV file</a></span></li></ul></div>

### 1. Import libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
%matplotlib inline
import scipy.stats
import seaborn as sns
sns.set(color_codes=True)

### 2. Load data into a dataframe

In [2]:
import os
working_directory = os.getcwd()
print(working_directory)

/Users/itunuolaitan/Documents/2022/datasets


In [3]:
path = working_directory + '/AB_NYC_2019.csv'
data = pd.read_csv(path)
data.head(10)

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   
5  5099         Large Cozy 1 BR Apartment In Midtown East     7322   
6  5121                                   BlissArtsSpace!     7356   
7  5178                  Large Furnished Room Near B'way      8967   
8  5203                Cozy Clean Guest Room - Family Apt     7490   
9  5238                Cute & Cozy Lower East Side 1 bdrm     7549   

     host_name neighbourhood_group       neighbourhood  latitude  longitude  \
0         John            Brooklyn          Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan             Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan              Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn        Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan         East Harlem  40.79851  -73.94399   
5        Chris           Manhattan         Murray Hill  40.74767  -73.97500   
6        Garon            Brooklyn  Bedford-Stuyvesant  40.68688  -73.95596   
7     Shunichi           Manhattan      Hell's Kitchen  40.76489  -73.98493   
8    MaryEllen           Manhattan     Upper West Side  40.80178  -73.96723   
9          Ben           Manhattan           Chinatown  40.71344  -73.99037   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaN   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   
5  Entire home/apt    200               3                 74  2019-06-22   
6     Private room     60              45                 49  2017-10-05   
7     Private room     79               2                430  2019-06-24   
8     Private room     79               2                118  2017-07-21   
9  Entire home/apt    150               1                160  2019-06-09   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0  
5               0.59                               1               129  
6               0.40                               1                 0  
7               3.47                               1               220  
8               0.99                               1                 0  
9               1.33                               4               188

Check many columns and rows the dataset holds

In [4]:
data.shape

(48895, 16)

### 3. Check datatypes within the data

In [5]:
data.dtypes

id                                  int64
name                               object
host_id                             int64
host_name                          object
neighbourhood_group                object
neighbourhood                      object
latitude                          float64
longitude                         float64
room_type                          object
price                               int64
minimum_nights                      int64
number_of_reviews                   int64
last_review                        object
reviews_per_month                 float64
calculated_host_listings_count      int64
availability_365                    int64
dtype: object

From the snapshot of the table above the following columns need to be corrected to the proper data type:

last_review: from `int64` to `datetime64` as this column holds time data

In [6]:
data['last_review'] = pd.to_datetime(data['last_review'], format='%Y-%m-%d')
data.head(5)

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaT   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

In [7]:
data.dtypes

id                                         int64
name                                      object
host_id                                    int64
host_name                                 object
neighbourhood_group                       object
neighbourhood                             object
latitude                                 float64
longitude                                float64
room_type                                 object
price                                      int64
minimum_nights                             int64
number_of_reviews                          int64
last_review                       datetime64[ns]
reviews_per_month                        float64
calculated_host_listings_count             int64
availability_365                           int64
dtype: object

The last_review column data type was converted from int64 to datetime64[ns].

### 4. Missing Values

Check how many missing values are in each of the columns

In [8]:
pd.DataFrame({"Missing values ": data.isnull().sum()})

Missing values 
id                                            0
name                                         16
host_id                                       0
host_name                                    21
neighbourhood_group                           0
neighbourhood                                 0
latitude                                      0
longitude                                     0
room_type                                     0
price                                         0
minimum_nights                                0
number_of_reviews                             0
last_review                               10052
reviews_per_month                         10052
calculated_host_listings_count                0
availability_365                              0

> name, host name, last_review and reviews_per_month have missing values in its columns

### 5. Remove NaN values from name and host_name

In [9]:
pd.DataFrame({"Missing values (%)" : data.isnull().sum()/len(data.index)*100})

Missing values (%)
id                                        0.000000
name                                      0.032723
host_id                                   0.000000
host_name                                 0.042949
neighbourhood_group                       0.000000
neighbourhood                             0.000000
latitude                                  0.000000
longitude                                 0.000000
room_type                                 0.000000
price                                     0.000000
minimum_nights                            0.000000
number_of_reviews                         0.000000
last_review                              20.558339
reviews_per_month                        20.558339
calculated_host_listings_count            0.000000
availability_365                          0.000000

Only a small percentage of missing values in the name and host_name columns (0.03% and 0.04% respectively) , so these NaN values will be removed. Will review the missing values in last_review and reviews_per_month columns as these contain 20% of missing data.

In [10]:
#remove missing values in name and host_name columns
original = len(data)
data = data.dropna(how='any', subset=['name', 'host_name'])
print('Number of NaN values removed:', original - len(data))

Number of NaN values removed: 37


### 6. Check for duplicate values

In [11]:
duplicate_rows_data = data[data.duplicated()]
print("number of duplicate rows: ", duplicate_rows_data.shape)

number of duplicate rows:  (0, 16)


There are no duplicate records in the data which means the all records are unique

### 7. Remove any invalid values 

In [12]:
print ('Number of minimum nights 0:', len(data[data['minimum_nights'] == 0]))
print ('Number of calculated host listing count 0:', len(data[data['calculated_host_listings_count'] == 0]))
print ('Number of days listing is available 0:', len(data[data['availability_365'] == 0]))
print ('Number of listings with price $0.00:', len(data[data['price'] == 0]))

Number of minimum nights 0: 0
Number of calculated host listing count 0: 0
Number of days listing is available 0: 17508
Number of listings with price $0.00: 11


> Within the dataset, there are 17508 records that have listings that are not available for a 365 day period. These are said to be listing that are **temporarily** unavailable due to host preferences and appeared to be set to 0 when the data was collected. As Airbnb provides this option to hosts and have not had listings taken down completely, data should be treated as relevant for exploratory descriptive analysis.

> There are 11 listings with prices listed as $0.00 - these are invalid records as there are no 'free' listings on Airbnb


In [13]:
#print the 11 records with $0.00 as listing price and investigate
data.loc[data['price'] == 0]

id                                               name    host_id  \
23161  18750597  Huge Brooklyn Brownstone Living, Close to it all.    8993084   
25433  20333471      ★Hostel Style Room | Ideal Traveling Buddies★  131697576   
25634  20523843    MARTIAL LOFT 3: REDEMPTION (upstairs, 2nd room)   15787004   
25753  20608117                    Sunny, Quiet Room in Greenpoint    1641537   
25778  20624541      Modern apartment in the heart of Williamsburg   10132166   
25794  20639628  Spacious comfortable master bedroom with nice ...   86327101   
25795  20639792  Contemporary bedroom in brownstone with nice view   86327101   
25796  20639914       Cozy yet spacious private brownstone bedroom   86327101   
26259  20933849                              the best you can find   13709292   
26841  21291569  Coliving in Brooklyn! Modern design / Shared room  101970559   
26866  21304320             Best Coliving space ever! Shared room.  101970559   

          host_name neighbourhood_group       neighbourhood  latitude  \
23161      Kimberly            Brooklyn  Bedford-Stuyvesant  40.69023   
25433        Anisha               Bronx     East Morrisania  40.83296   
25634  Martial Loft            Brooklyn            Bushwick  40.69467   
25753        Lauren            Brooklyn          Greenpoint  40.72462   
25778       Aymeric            Brooklyn        Williamsburg  40.70838   
25794       Adeyemi            Brooklyn  Bedford-Stuyvesant  40.68173   
25795       Adeyemi            Brooklyn  Bedford-Stuyvesant  40.68279   
25796       Adeyemi            Brooklyn  Bedford-Stuyvesant  40.68258   
26259        Qiuchi           Manhattan         Murray Hill  40.75091   
26841        Sergii            Brooklyn            Bushwick  40.69211   
26866        Sergii            Brooklyn            Bushwick  40.69166   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
23161  -73.95428     Private room      0               4                  1   
25433  -73.88668     Private room      0               2                 55   
25634  -73.92433     Private room      0               2                 16   
25753  -73.94072     Private room      0               2                 12   
25778  -73.94645  Entire home/apt      0               5                  3   
25794  -73.91342     Private room      0               1                 93   
25795  -73.91170     Private room      0               1                 95   
25796  -73.91284     Private room      0               1                 95   
26259  -73.97597  Entire home/apt      0               3                  0   
26841  -73.90670      Shared room      0              30                  2   
26866  -73.90928      Shared room      0              30                  5   

      last_review  reviews_per_month  calculated_host_listings_count  \
23161  2018-01-06               0.05                               4   
25433  2019-06-24               2.56                               4   
25634  2019-05-18               0.71                               5   
25753  2017-10-27               0.53                               2   
25778  2018-01-02               0.15                               1   
25794  2019-06-15               4.28                               6   
25795  2019-06-21               4.37                               6   
25796  2019-06-23               4.35                               6   
26259         NaT                NaN                               1   
26841  2019-06-22               0.11                               6   
26866  2019-05-24               0.26                               6   

       availability_365  
23161                28  
25433               127  
25634                 0  
25753                 0  
25778                73  
25794               176  
25795               232  
25796               222  
26259                 0  
26841               333  
26866               139

As all other information on the $0.00 listings are valid, these records will be dropped

In [14]:
original = len(data)
data = data[data.price !=0]
print('Number of listing prices as $0.00 removed:', original - len(data))

Number of listing prices as $0.00 removed: 11


### 8. Convert NaN reviews per month and NaT last reviews into 'No Reviews'

Where the number_of_reviews = 0, I will convert the NaN values in the `reviews_per_month` column and the NaT values in the `last_review` column into 'No Reviews' for ease of readibility. Any null values that are left once the criteria has been applied, will be removed

In [15]:
#replace the NaN reviews_per_month with 'No Reviews'
values = data['reviews_per_month'][data['number_of_reviews'] == 0].index.values.tolist()
data.loc[values, 'reviews_per_month'] = data['reviews_per_month'][data['number_of_reviews'] == 0].replace(np.nan, 'No Reviews')

In [16]:
#replace the NaT last_review with 'No Reviews'
values = data['last_review'][data['number_of_reviews'] == 0].index.values.tolist()
data.loc[values, 'last_review'] = data['last_review'][data['number_of_reviews'] == 0].replace(np.nan, 'No Reviews')

In [17]:
data.head(5)

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews  \
0     Private room    149               1                  9   
1  Entire home/apt    225               1                 45   
2     Private room    150               3                  0   
3  Entire home/apt     89               1                270   
4  Entire home/apt     80              10                  9   

           last_review reviews_per_month  calculated_host_listings_count  \
0  2018-10-19 00:00:00              0.21                               6   
1  2019-05-21 00:00:00              0.38                               2   
2           No Reviews        No Reviews                               1   
3  2019-07-05 00:00:00              4.64                               1   
4  2018-11-19 00:00:00               0.1                               1   

   availability_365  
0               365  
1               355  
2               365  
3               194  
4                 0

Check if there are any missing values after conversions

In [18]:
pd.DataFrame({"Missing values (%)" : data.isnull().sum()/len(data.index)*100})

Missing values (%)
id                                             0.0
name                                           0.0
host_id                                        0.0
host_name                                      0.0
neighbourhood_group                            0.0
neighbourhood                                  0.0
latitude                                       0.0
longitude                                      0.0
room_type                                      0.0
price                                          0.0
minimum_nights                                 0.0
number_of_reviews                              0.0
last_review                                    0.0
reviews_per_month                              0.0
calculated_host_listings_count                 0.0
availability_365                               0.0

Results above show that are no missing values in the dataset

### 9. Convert Latitude and Longitude to 5 decimal places

In [19]:
data['latitude'].round(decimals = 5)

0        40.64749
1        40.75362
2        40.80902
3        40.68514
4        40.79851
           ...   
48890    40.67853
48891    40.70184
48892    40.81475
48893    40.75751
48894    40.76404
Name: latitude, Length: 48847, dtype: float64

In [20]:
data['longitude'].round(decimals = 5)

0       -73.97237
1       -73.98377
2       -73.94190
3       -73.95976
4       -73.94399
           ...   
48890   -73.94995
48891   -73.93317
48892   -73.94867
48893   -73.99112
48894   -73.98933
Name: longitude, Length: 48847, dtype: float64

### 10. Read dataframe to CSV file

In [21]:
data.to_csv('AB_NYC_2019_Clean.csv', index=False)